README.md
Project Structure
--------

  ```sh
  ├── Procfile
  ├── Procfile.dev
  ├── README.md
  ├── app.py
  ├── config.py
  ├── error.log
  ├── forms.py
  ├── models.py
  ├── requirements.txt
  ├── static
  │   ├── css
  │   │   ├── bootstrap-3.0.0.min.css
  │   │   ├── bootstrap-theme-3.0.0.css
  │   │   ├── bootstrap-theme-3.0.0.min.css
  │   │   ├── font-awesome-3.2.1.min.css
  │   │   ├── layout.forms.css
  │   │   ├── layout.main.css
  │   │   ├── main.css
  │   │   ├── main.quickfix.css
  │   │   └── main.responsive.css
  │   ├── font
  │   │   ├── FontAwesome.otf
  │   │   ├── fontawesome-webfont.eot
  │   │   ├── fontawesome-webfont.svg
  │   │   ├── fontawesome-webfont.ttf
  │   │   └── fontawesome-webfont.woff
  │   ├── ico
  │   │   ├── apple-touch-icon-114-precomposed.png
  │   │   ├── apple-touch-icon-144-precomposed.png
  │   │   ├── apple-touch-icon-57-precomposed.png
  │   │   ├── apple-touch-icon-72-precomposed.png
  │   │   └── favicon.png
  │   ├── img
  │   └── js
  │       ├── libs
  │       │   ├── bootstrap-3.0.0.min.js
  │       │   ├── jquery-1.10.2.min.js
  │       │   ├── modernizr-2.6.2.min.js
  │       │   └── respond-1.3.0.min.js
  │       ├── plugins.js
  │       └── script.js
  └── templates
      ├── errors
      │   ├── 404.html
      │   └── 500.html
      ├── forms
      │   ├── forgot.html
      │   ├── login.html
      │   └── register.html
      ├── layouts
      │   ├── form.html
      │   └── main.html
      └── pages
          ├── placeholder.about.html
          └── placeholder.home.html
  ```

### Quick Start

1. source env/bin/activate

### URL

1. fa16-cs411-18.cs.illinois.edu

changes.txt
from sqlalchemy import create_engine, scoped_session, sessionmaker

Session = scoped_session( sessionmaker() )
engine = create_engine('mysql+mysqldb://scott:tiger@localhost/foo')

app.config['SQLALCHEMY_POOL_SIZE'] = 100
app.config['SQLALCHEMY_POOL_RECYCLE'] = 280


## IN EACH FUNCTION
session = Session()
result = session.execute( someMappedClass )
# use session like normal ...
session.close()
# RETURN



http://stackoverflow.com/questions/6471549/avoiding-mysql-server-has-gone-away-on-infrequently-used-python-flask-server
http://stackoverflow.com/questions/17972020/how-to-execute-raw-sql-in-sqlalchemy-flask-app
http://docs.sqlalchemy.org/en/latest/orm/session_api.html#sqlalchemy.orm.session.Session


forms.py
from flask_wtf import Form
from wtforms import TextField, PasswordField
from wtforms.validators import DataRequired, EqualTo, Length

# Set your classes here.

class SearchForm(Form):
    search_term = TextField('Search', validators=[DataRequired(), Length(min=3, max=50)])

class RegisterForm(Form):
    name = TextField(
        'Username', validators=[DataRequired(), Length(min=6, max=25)]
    )
    email = TextField(
        'Email', validators=[DataRequired(), Length(min=6, max=40)]
    )
    password = PasswordField(
        'Password', validators=[DataRequired(), Length(min=6, max=40)]
    )
    confirm = PasswordField(
        'Repeat Password',
        [DataRequired(),
        EqualTo('password', message='Passwords must match')]
    )


class LoginForm(Form):
    name = TextField('Username', [DataRequired()])
    password = PasswordField('Password', [DataRequired()])


class ForgotForm(Form):
    email = TextField(
        'Email', validators=[DataRequired(), Length(min=6, max=40)]
    )

class FavoriteForm(Form):
    pass


app.py
#----------------------------------------------------------------------------#
# Imports
#----------------------------------------------------------------------------#

from flask import Flask, render_template, request
# from flask.ext.sqlalchemy import SQLAlchemy
import logging
from logging import Formatter, FileHandler
from forms import *
from sqlalchemy import create_engine
from sqlalchemy.orm import scoped_session, sessionmaker
import os
import MySQLdb
import urllib

#----------------------------------------------------------------------------#
# App Config.
#----------------------------------------------------------------------------#

app = Flask(__name__)
app.config.from_object('config')

Session = scoped_session( sessionmaker() )
engine = create_engine('mysql+mysqldb://root:cs411fa2016@localhost/actor_pages')

app.config['SQLALCHEMY_POOL_SIZE'] = 100
app.config['SQLALCHEMY_POOL_RECYCLE'] = 280

#db = SQLAlchemy(app)
db = MySQLdb.connect(host='localhost', user='root', passwd='cs411fa2016', db='actor_pages')
# Automatically tear down SQLAlchemy.

'''
@app.teardown_request
def shutdown_session(exception=None):
    db.close() 
'''

# Login required decorator.
'''
def login_required(test):
    @wraps(test)
    def wrap(*args, **kwargs):
        if 'logged_in' in session:
            return test(*args, **kwargs)
        else:
            flash('You need to login first.')
            return redirect(url_for('login'))
    return wrap
'''
#----------------------------------------------------------------------------#
# Controllers.
#----------------------------------------------------------------------------#


@app.route('/')
def home():
    cur = db.cursor()
    cur.execute('SELECT Actor.name,Actor.actorID FROM Actor, Favorites WHERE Favorites.userID = 1 AND Actor.actorID = Favorites.actorID ORDER BY Favorites.rank;')
    results = [ (name,urllib.quote(str(actor_id))) for (name,actor_id) in cur.fetchall() ]
    return render_template('pages/home.html', res=results)

@app.route('/actors', methods=['GET', 'POST'])
def actors():
    form = SearchForm()
    if form.validate_on_submit():
        # Split name at commas and spaces
        comma_split = form.search_term.data.split(',')
        search_terms = []
        for term in comma_split:
		search_terms = search_terms + [x.strip() for x in term.split(' ')]
 
        # Build query from each part of name
        query = 'SELECT actorID, name FROM Actor WHERE name LIKE \'%' + search_terms[0] + '%\''
        for i in range(1,len(search_terms)):
            query += 'AND name LIKE \'%' + search_terms[i] + '%\''
        query += ';'

        cur = db.cursor()
        cur.execute(query)
        actor_list = [ (urllib.quote(str(actor_id)), name) for (actor_id, name) in cur.fetchall()[0:20] ]

        return render_template('pages/actors.html', form=form, actor_list=actor_list)
    return render_template('pages/actors.html', form=form, actor_list=[])

@app.route('/actor/<actor_id>', methods=['GET', 'POST'])
def actor(actor_id):
    form = FavoriteForm()
    query = 'SELECT name FROM Actor WHERE actorID=' + actor_id + ';'
    cur = db.cursor()
    cur.execute(query)
    actor_name = cur.fetchone()
    movie_query = 'SELECT m.MovieID, m.title, m.rating, m.year FROM Movies m, ActsIn a WHERE m.MovieID=a.MovieID AND a.ActorID=' + actor_id + ';'
    cur = db.cursor()
    cur.execute(movie_query)
    movies = [ (movieID, unicode(str(title), 'latin-1'), urllib.quote(str(rating)), urllib.quote(str(year))) for (movieID, title, rating, year) in cur.fetchall() ]
    movie_genres = {}
    for movieID,title,rating,year in movies:
        genre_query = 'SELECT g.genreName FROM Genre g, HasGenre hg WHERE g.genreID=hg.genreID AND hg.MovieID=' + str(movieID) +';'
        cur = db.cursor()
        cur.execute(genre_query)
        movie_genres[(title,rating,year)] = [str(genre).strip('(').strip('),').strip('\'') for genre in cur.fetchall()]
    if form.validate_on_submit():
        user_id=1
        favorite_query = 'DELETE FROM Favorites WHERE userID=' + str(user_id) + ' AND (rank=10 OR actorID=' + str(actor_id) + ');'
        cur = db.cursor()
        cur.execute(favorite_query)
        update_query = 'UPDATE Favorites SET rank=rank+1 WHERE 1;'
        cur = db.cursor()
        cur.execute(update_query)
        add_favorite_query = 'INSERT INTO Favorites VALUES(' + str(user_id) + ',' + actor_id + ',1);'
        cur = db.cursor()
        cur.execute(add_favorite_query) 
        return render_template('pages/actor.html', add_favorite_form=form, actor_id=actor_id, actor_name=actor_name, movies=movie_genres)
    return render_template('pages/actor.html', add_favorite_form=form, actor_id=actor_id, actor_name=actor_name, movies=movie_genres)

@app.route('/about')
def about():
    return render_template('pages/about.html')


@app.route('/login')
def login():
    form = LoginForm(request.form)
    return render_template('forms/login.html', form=form)


@app.route('/register')
def register():
    form = RegisterForm(request.form)
    return render_template('forms/register.html', form=form)


@app.route('/forgot')
def forgot():
    form = ForgotForm(request.form)
    return render_template('forms/forgot.html', form=form)

# Error handlers.


@app.errorhandler(500)
def internal_error(error):
    #db_session.rollback()
    return render_template('errors/500.html'), 500


@app.errorhandler(404)
def not_found_error(error):
    return render_template('errors/404.html'), 404

if not app.debug:
    file_handler = FileHandler('error.log')
    file_handler.setFormatter(
        Formatter('%(asctime)s %(levelname)s: %(message)s [in %(pathname)s:%(lineno)d]')
    )
    app.logger.setLevel(logging.INFO)
    file_handler.setLevel(logging.INFO)
    app.logger.addHandler(file_handler)
    app.logger.info('errors')

#----------------------------------------------------------------------------#
# Launch.
#----------------------------------------------------------------------------#

# Default port:
'''
if __name__ == '__main__':
    app.run()
'''
# Or specify port manually:

if __name__ == '__main__':
    port = int(os.environ.get('PORT', 5000))
    app.run(host='0.0.0.0', port=port)

